### PR TITLE
http: remove getAll() header map API and switch all usages to get()

### PR DESCRIPTION
**Pull request by mattklein123**
This is a follow up to https://github.com/envoyproxy/envoy/commit/2c60632d41555ec8b3d9ef5246242be637a2db0f.
This forces all callers to think about multiple header values. There may be places that we want
to support multiple values, but none of them are security critical and this change should be
functionally equivalent to what exists today.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@alyssawilk lmk if you want a comment for every converted prod code site. I think I had that in setec at some point and can bring it back.

**Comment by alyssawilk** [comment]
yeah, I think any point we're automatically pulling [0] we should comment why that's OK.

**Comment by mattklein123** [comment]
OK I will update.

File: source/extensions/filters/common/fault/fault_config.cc [hunk 1]
```diff
@@ -18,12 +18,13 @@ HeaderPercentageProvider::percentage(const Http::RequestHeaderMap* request_heade
...
   }
   const auto header = request_headers->get(header_name_);
-  if (header == nullptr) {
+  if (header.empty()) {
     return percentage_;
   }
 
   uint32_t header_numerator;
-  if (!absl::SimpleAtoi(header->value().getStringView(), &header_numerator)) {
+  // Client controlled. Using the first value is fine.
```
Comment (alyssawilk) [inline source/extensions/filters/common/fault/fault_config.cc:26]: (belatedly) I'm not sure this comment is going to be illuminating to someone reading the code.
Maybe something along the lines of "This header is cleared unless set by trusted clients, and they should not be sending multi-lined values"?
for the "client controlled" ones here, above and below?

File: source/extensions/filters/http/cache/cache_headers_utils.cc [hunk 2]
```diff
@@ -271,8 +272,8 @@ bool VaryHeader::isAllowed(const Http::ResponseHeaderMap& headers) const {
 }
 
 bool VaryHeader::hasVary(const Http::ResponseHeaderMap& headers) {
-  const Http::HeaderEntry* vary_header = headers.get(Http::Headers::get().Vary);
-  return vary_header != nullptr && !vary_header->value().empty();
+  const auto vary_header = headers.get(Http::Headers::get().Vary);
+  return !vary_header.empty() && !vary_header[0]->value().empty();
```
Comment (alyssawilk) [inline source/extensions/filters/http/cache/cache_headers_utils.cc:276]: so vary can definitely be multi-value.  While it's not reasonable to send
Vary:
Var: foo, bar
I think it's legal.

I wonder if we can just remove this utility and do a quick vary_header.empty() check in isAllowed for the performance wins?

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 3]
```diff
@@ -183,7 +183,7 @@ void SimpleHttpCache::varyInsert(const Key& request_key,
   if (iter == map_.end()) {
     Http::ResponseHeaderMapPtr vary_only_map =
         Http::createHeaderMap<Http::ResponseHeaderMapImpl>({});
-    vary_only_map->setCopy(Http::Headers::get().Vary, vary_header->value().getStringView());
+    vary_only_map->setCopy(Http::Headers::get().Vary, vary_header[0]->value().getStringView());
```
Comment (alyssawilk) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:186]: shouldn't this support multi-value headers?
Maybe a TODO for caching folks to look into this one?

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 4]
```diff
@@ -200,7 +200,7 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
       // Currently it is not possible to add {{"append": "1"}, {"append": "2"}} (the intended
       // combined headers: {{"original": "true"}, {"append": "1"}, {"append": "2"}}) to the request
       // to upstream server by only sets `headers_to_append`.
-      if (header_to_modify != nullptr) {
+      if (!header_to_modify.empty()) {
```
Comment (alyssawilk) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:203]: In general I wonder if it's worth a docs comment somewhere that header to append appends to first value.
If had a tracing header in a non-envoy server which did multi-value,
Trace: ServerA
Trace: ServerB
I think Envoy as server C would generally end up
Trace: ServerA, ServerC
Trace: ServerB.
We offer no header order guarantees so really anyone who cares about order is ideally doing inline appends anyway, so optional.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc [hunk 5]
```diff
@@ -63,9 +63,9 @@ RouteConstSharedPtr RouteEntryImplBase::clusterEntry(uint64_t random_value,
   const auto& cluster_header = clusterHeader();
   if (!cluster_header.get().empty()) {
     const auto& headers = metadata.headers();
-    const auto* entry = headers.get(cluster_header);
-    if (entry != nullptr) {
-      return std::make_shared<DynamicRouteEntry>(*this, entry->value().getStringView());
+    const auto entry = headers.get(cluster_header);
+    if (!entry.empty()) {
```
Comment (alyssawilk) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:67]: comment here and below?

**Review by alyssawilk — COMMENTED** [review]
Awesome, thanks for the comments.
Sending some comments back your way, mostly optional but a few I think worth cleaning up for clarity

**Comment by mattklein123** [comment]
Once we agree on this PR I'm going to need to update envoy-filter-example to get this to pass FYI.

File: source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc [hunk 6]
```diff
@@ -136,9 +137,10 @@ void HttpGrpcAccessLog::emitLog(const Http::RequestHeaderMap& request_headers,
     auto* logged_headers = response_properties->mutable_response_headers();
 
     for (const auto& header : response_headers_to_log_) {
-      const Http::HeaderEntry* entry = response_headers.get(header);
-      if (entry != nullptr) {
-        logged_headers->insert({header.get(), std::string(entry->value().getStringView())});
+      const auto entry = response_headers.get(header);
+      if (!entry.empty()) {
+        // TODO(mattklein123): Potentially log all header values.
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc:142]: https://github.com/envoyproxy/envoy/issues/13454

File: source/extensions/filters/http/cache/cache_headers_utils.cc (hunk 2 above)
Comment (mattklein123) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:276]: Will put a TODO in here for caching people to look at.

Comment (mattklein123) [reply source/extensions/filters/http/cache/cache_headers_utils.cc:276]: Now that I look at this code more, I think a lot of this code would be vastly simpler if Vary where just a custom inline header, but that is getting into a larger change than I want to make, so I will just put that into the TODO.

File: source/extensions/filters/http/ext_authz/ext_authz.cc (hunk 4 above)
Comment (mattklein123) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:203]: Honestly I find this code mind boggingly confusing every time I look at this, between gRPC/HTTP/etc. @dio do you mind taking a look and seeing if we can improve the docs here for extauth about this behavior?

**Comment by mattklein123** [comment]
@alyssawilk I think I covered most of the comments. PTAL. When you are satisfied I will fix envoy-filter-example and we can merge in lockstep.

**Comment by mattklein123** [comment]
See https://github.com/envoyproxy/envoy-filter-example/pull/135 for required filter example fix.

**Review by alyssawilk — DISMISSED** [review]
Awesome!  I'm going to let you merge this and the filter example at your convenience :-)

**Comment by alyssawilk** [comment]
Hm, looks like CI needs a pass and it needs a merge though.  I'll do another pass once you have that sorted.

**Review by alyssawilk — APPROVED** [review]

File: source/extensions/common/wasm/context.cc [hunk 7]
```diff
@@ -696,16 +696,18 @@ WasmResult Context::getHeaderMapValue(WasmHeaderMapType type, absl::string_view
...
     if (wasm()->abiVersion() == proxy_wasm::AbiVersion::ProxyWasm_0_1_0) {
       *value = "";
       return WasmResult::Ok;
     } else {
       return WasmResult::NotFound;
     }
   }
-  *value = entry->value().getStringView();
+  // TODO(kyessenov, PiotrSikora): This needs to either return a concatenated list of values, or
+  // the ABI needs to be changed to return multiple values. This is a potential security issue.
```
Comment (mattklein123) [inline source/extensions/common/wasm/context.cc:709]: @kyessenov @PiotrSikora this needs to be fixed ASAP

**Comment by mattklein123** [comment]
@alyssawilk I need another approve if you have a sec.

File: source/extensions/common/wasm/context.cc (hunk 7 above)
Comment (mattklein123) [reply source/extensions/common/wasm/context.cc:709]: (Talked to @kyessenov offline about this and it will be fixed soon)

**Review by alyssawilk — APPROVED** [review]

File: source/extensions/common/wasm/context.cc (hunk 7 above)
Comment (PiotrSikora) [reply source/extensions/common/wasm/context.cc:709]: FWIW, the ABI in proxy-wasm/spec#1 already returns multiple values.

Comment (mattklein123) [reply source/extensions/common/wasm/context.cc:709]: Perfect. Let's do that then.

**Pull request merged**